src/assembler/rvassembler_common.h

```diff
@@ -6,49 +6,208 @@
+template <typename Reg_T>
+class RVOpcode : public Opcode<Reg_T> {
+public:
+  RVOpcode(const Token &name, RVISA::Opcode opcode)
+      : Opcode<Reg_T>(name, {RVOpPartOpcode(opcode)}) {}
+
+  RVOpcode(const Token &name, RVISA::Opcode opcode, RVOpPartFunct3 funct3)
+      : Opcode<Reg_T>(name, {RVOpPartOpcode(opcode), funct3}) {}
+
+  RVOpcode(const Token &name, RVISA::Opcode opcode, RVOpPartFunct3 funct3,
+           RVOpPartFunct7 funct7)
+      : Opcode<Reg_T>(name, {RVOpPartOpcode(opcode), funct3, funct7}) {}
+};
+
+template <typename Reg_T>
+class RVRegRs1 : public Reg<Reg_T> {
+public:
+  RVRegRs1(const ISAInfoBase *isa, unsigned fieldIndex)
+      : Reg<Reg_T>(isa, fieldIndex, 15, 19, "rs1") {}
+};
+
+template <typename Reg_T>
+class RVRegRs2 : public Reg<Reg_T> {
+public:
+  RVRegRs2(const ISAInfoBase *isa, unsigned fieldIndex)
+      : Reg<Reg_T>(isa, fieldIndex, 20, 24, "rs2") {}
+};
+
+template <typename Reg_T>
+class RVRegRd : public Reg<Reg_T> {
+public:
+  RVRegRd(const ISAInfoBase *isa, unsigned fieldIndex)
+      : Reg<Reg_T>(isa, fieldIndex, 7, 11, "rd") {}
+};
+
+template <typename Reg_T, class RVReg_T>
+std::shared_ptr<Reg<Reg_T>> makeRvReg(const ISAInfoBase *isa, unsigned fieldIndex) {
+  return std::make_shared<Reg<Reg_T>>(RVReg_T(isa, fieldIndex));
+}
+
+template <typename Reg_T>
+class RVImm5 : public Imm<Reg_T> {
+public:
+  RVImm5(unsigned fieldIndex)
+      : Imm<Reg_T>(fieldIndex, 5, Imm<Reg_T>::Repr::Unsigned,
+                   std::vector{ImmPart(0, 20, 24)}) {}
+};
+
+template <typename Reg_T>
+class RVImm12 : public Imm<Reg_T> {
+public:
+  RVImm12(unsigned fieldIndex)
+      : Imm<Reg_T>(fieldIndex, 12, Imm<Reg_T>::Repr::Signed,
+                   std::vector{ImmPart(0, 20, 31)}) {}
+};
+
+template <typename Reg_T>
+class RVImm13 : public Imm<Reg_T> {
+public:
+  RVImm13(unsigned fieldIndex)
+      : Imm<Reg_T>(fieldIndex, 13, Imm<Reg_T>::Repr::Signed,
+                   std::vector{ImmPart(12, 31, 31), ImmPart(11, 7, 7),
+                               ImmPart(5, 25, 30), ImmPart(1, 8, 11)},
+                   Imm<Reg_T>::SymbolType::Relative) {}
+};
+
+// A B-Type RISC-V instruction
+template <typename Reg_T>
+class BTypeInstr : public RVInstruction<Reg_T> {
+public:
+  BTypeInstr(const Token &name, unsigned funct3, const ISAInfoBase *isa)
+      : RVInstruction<Reg_T>(
+            Opcode<Reg_T>(name, {RVOpPartOpcode(RVISA::Opcode::BRANCH),
+                          RVOpPartFunct3(funct3)}),
+            {makeRvReg<Reg_T, RVRegRs1<Reg_T>>(isa, 1),
+             makeRvReg<Reg_T, RVRegRs2<Reg_T>>(isa, 2),
+             std::make_shared<Imm<Reg_T>>(RVImm13<Reg_T>(3))
+            }) {}
+};
+
+template <typename Reg_T>
+std::shared_ptr<Instruction<Reg_T>> defineBType(const QString &name, unsigned funct3,
+                                                       const ISAInfoBase *isa) {
+  return std::shared_ptr<Instruction<Reg_T>>(
+      new BTypeInstr<Reg_T>(Token(name), funct3, isa));
+}
```

Comment:
If the templates are this small, then we might aswell just call `std::make_shared` at the callsite, instead of these `define` functions.